<html>
<head>
    <title>Annotations, heatmap | Ideogram</title>
    <script type="text/javascript" src="../../dist/js/ideogram.min.js"></script>
    <style>
        body {
            font: 12px Arial;
            line-height: 19.6px;
        }

        ul {
            list-style: none;
            padding-left: 10px;
            float: left;
        }

        .note {
            font-style: italic;
            padding-left: 10px;
            clear: left;
        }
    </style>
</head>
<body>
<h1>Annotations, heatmap | Ideogram</h1>
<a href="../">Back to overview</a> |
<a href="annotations-file-url">Previous</a> |
<a href="annotations-animated">Next</a>
<br/><br/>
<div>
    This ideogram shows the distribution of all human genes throughout the genome,
    by expression level.
</div>
<div id="container"></div>

<div class="note">
    Expression data from NCBI SRA and GEO (<a href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR562645">SRR562645</a>, <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE40705">GSE40705</a>)
    via <a href="https://github.com/NCBI-Hackathons/rnaseqview#rnaseqview">rnaseqview</a> pipeline.
</div>
<script type="text/javascript">

  function colorAnnotsByExpression() {

    var t0 = new Date().getTime();

    var raContainer, chr, rawAnnots, ra, i, j, trackIndex,
      newRaContainers, newRa, newRas, expressionValue, color,
      geneType,
      keys = ideogram.rawAnnots.keys,
      rawAnnotContainers = ideogram.rawAnnots.annots;

    newRaContainers = [];

    for (i = 0; i < rawAnnotContainers.length; i++) {

      raContainer = rawAnnotContainers[i];
      chr = raContainer.chr;
      rawAnnots = raContainer.annots;
      newRas = [];

      for (j = 0; j < rawAnnots.length; j++) {

        ra = rawAnnots[j]; // [name, start, length, expression-value, gene-type]
        expressionValue = ra[3];

        newRa = ra.slice(0, 3); // name, start, length
        if (expressionValue > 3) {
          color = '#F33'; // red
        } else if (expressionValue > 0 && expressionValue <= 3) {
          color = '#88F'; // blue
        } else {
          color = '#AAA'; // grey
        }
        trackIndex = 0;
        newRa.push(trackIndex, color, expressionValue);
        newRas.push(newRa);

        geneType = ra[4];
        newRa = ra.slice(0, 3);
        switch (geneType) {
          case 4:
            color = '#F00';
            break;
          case 3:
            color = '#FA0';
            break;
          case 2:
            color = '#AAA';
            break;
          case 1:
            color = '#0AF';
            break;
          case 0:
            color = '#00F';
            break;
        }
        trackIndex = 1;
        newRa.push(trackIndex, color, geneType);
//        console.log(newRa)
        newRas.push(newRa);

      }
      newRaContainers.push({chr: chr, annots: newRas});
    }

    keys.splice(3, 0, 'trackIndex');
    keys.splice(4, 0, 'color');

    ideogram.rawAnnots.keys = keys;
    ideogram.rawAnnots.annots = newRaContainers;

    var t1 = new Date().getTime();
    if (ideogram.config.debug) {
      console.log('Time in colorAnnotsByExpression: ' + (t1 - t0) + ' ms');
    }
  }

  var annotationTracks = [
    {id: 'expressionLevelTrack', displayName: 'Expression level'},
    {id: 'geneTypeTrack', displayName: 'Gene type'},
  ];

  var config = {
    container: '#container',
    organism: 'human',
    assembly: 'GRCh37',
    chrHeight: 275,
    annotationsPath: '../../data/annotations/SRR562646.json',
    annotationsLayout: 'heatmap',
    annotationTracks: annotationTracks,
    onLoadAnnots: colorAnnotsByExpression
  };

  var ideogram = new Ideogram(config);
</script>
</body>
</html>
