<!DOCTYPE html>
<html>
<head>
  <title>Related genes | Ideogram</title>
  <style>
    body {font: 14px Arial; line-height: 19.6px; padding: 0 15px;}
    a, a:visited {text-decoration: none;}
    a:hover {text-decoration: underline;}
    a, a:hover, a:visited, a:active {color: #0366d6;}

    select, optgroup, option {
      float: left;
      font-size: 14px;
    }

    #ideogram-container {z-index: -1}

    #search-container {
      height: 26px;
      float: left;
      position: relative;
      top: -4px;
      margin-left: 10px;
    }

    #search-genes {
      padding-left: 3px;
      width: 200px;
      height: 20px;
      font-size: 14px;
      border-radius: 3px;
      border: 1px solid #888;
    }

    #search-button {
      height: 21px;
      width: 23px;
      font-size: 24px;
      background: #58F;
      color: #FFF;
      display: inline-block;
      position: relative;
      top: 3px;
      left: -25px;
      border-radius: 3px;
      text-align: center;
      padding-top: 3px;
      cursor: pointer;
    }

    #ideogram .annot path {
      cursor: pointer;
    }
  </style>
  <script type="text/javascript" src="../../dist/js/ideogram.min.js"></script>
<link rel="icon" type="image/x-icon" href="img/ideogram_favicon.ico">
</head>
<body>
  <h1>Related genes | Ideogram</h1>
  <a href="../">Overview</a> |
  <a href="annotations-heatmap">Previous</a> |
  <a href="differential-expression">Next</a> |
  <a href="https://github.com/eweitz/ideogram/blob/gh-pages/related-genes.html" target="_blank">Source</a>
  <p>
    Find interacting pathway genes and paralogs.  Uses data from
    <a href="https://rest.ensembl.org">Ensembl</a>,
    <a href="https://www.wikipathways.org">WikiPathways</a>, and
    <a href="https://mygene.info">MyGene.info</a>.
    See also <a href="orthologs">Orthologs</a>.
  </p>
  <div>
    <select>
      <optgroup label="Model organisms">
        <option value="homo-sapiens" selected>Human (Homo sapiens)</option>
        <option value="mus-musculus">Mouse (Mus musculus)</option>
        <option value="rattus-norvegicus">Rat (Rattus norvegicus)</option>
        <!-- <option value="drosophila-melanogaster">Fly (Drosophila melanogaster)</option> -->
        <!-- <option value="caenorhabditis-elegans">Worm (Caenorhabditis elegans)</option> -->
        <option value="danio-rerio">Zebrafish (Danio rerio)</option>
        <!-- <option value="arabidopsis-thaliana">Thale cress (Arabidopsis thaliana)</option> -->
        <!-- <option value="saccharomyces-cerevisiae">Yeast (Saccharomyces cerevisiae)</option> -->
      </optgroup>
      <optgroup label="Vertebrates">
        <option value="pan-troglodytes">Chimpanzee (Pan troglodytes)</option>
        <option value="macaca-mulatta">Macaque (Macaca mulatta)</option>
        <option value="macaca-fascicularis">Crab-eating macaque (Macaca fascicularis)</option>
        <option value="felis-catus">Cat (Felis catus)</option>
        <option value="canis-lupus-familiaris">Dog (Canis lupus familiaris)</option>
        <option value="equus-caballus">Horse (Equus caballus)</option>
        <option value="bos-taurus">Cow (Bos taurus)</option>
        <option value="sus-scrofa">Pig (Sus scrofa)</option>
        <!-- <option value="petromyzon-marinus">Lamprey (Petromyzon marinus)</option> -->
      </optgroup>
      <!-- <optgroup label="Plants">
        <option value="zea-mays">Maize (Zea mays)</option>
        <option value="oryza-sativa">Rice (Oryza sativa)</option>
        <option value="solanum-lycopersicum">Tomato (Solanum lycopersicum)</option>
        <option value="musa-acuminata">Banana (Musa acuminata)</option>
        <option value="vitis-vinifera">Grape (Vitis vinifera)</option>
        <option value="micromonas-commoda">Green algae (Micromonas commoda)</option>
      </optgroup>
      <optgroup label="Insects">
        <option value="anopheles-gambiae">Mosquito (Anopheles gambiae)</option>
      </optgroup>
      <optgroup label="Protozoa">
        <option value="plasmodium falciparum">Malaria parasite (Plasmodium falciparum)</option>
        <option value="leishmania donovani">Leishmania parasite (Leishmania donovani)</option>
      </optgroup> -->
    </select>
  </div>
  <div style="float: left; width: 350px;">
    <label for="search-genes" id="search-container">
      <input id="search-genes" autocomplete="off" placeholder="Search gene (e.g. RAD51)"/>
      <span id="search-button">&#x2315;</span>
    </label>
  </div>
  <br/><br/><br/>
  <div id="ideogram-container"></div>
  <script type="text/javascript">

    ////
    // Start example-specific handling
    ///

    const urlParams = parseUrlParams();

    const organism = 'org' in urlParams ? urlParams.org : 'homo-sapiens';

    document.querySelector(`[value="${organism}"]`).selected = true;
    document.querySelector('select').addEventListener('change', handleOrganism);

    if ('q' in urlParams) {
      document.querySelector('#search-genes').value = urlParams['q']
    }

    document.querySelector('#search-button').addEventListener('click', handleSearch);
    document.querySelector('#search-genes').addEventListener('keyup', handleSearch);

    function parseUrlParams() {
      let rawParams = document.location.search;
      const urlParams = {};
      var param, key, value;
      if (rawParams !== '') {
        rawParams = rawParams.split('?')[1].split('&');
        rawParams.forEach(rawParam => {
          param = rawParam.split('=');
          key = param[0];
          value = param[1];
          urlParams[key] = value;
        });
      }
      return urlParams;
    }

    // Record app state in URL
    function updateUrl() {
      const params = Object.keys(urlParams).map(key => {
        return key + '=' + urlParams[key];
      }).join('&');
      history.pushState(null, null, '?' + params);
    }

    function handleOrganism() {
      const selectedOrg =
        document.querySelector('option:checked').text
          .split('(')[1].split(')')[0]
          .replace(/ /g, '-').toLowerCase();

      urlParams['org'] = selectedOrg;
      updateUrl();

      config.organism = selectedOrg;
      // const banded = !['homo sapiens', 'mus musculus'].includes(selectedOrg);
      // config.showFullyBanded = banded;
      delete ideogram;
      ideogram = new Ideogram(config);

    }

    // Process text input for the "Search" field.
    function handleSearch(event) {
      // Ignore non-"Enter" keyups
      if (event.type === 'keyup' && event.keyCode !== 13) return;

      const searchInput = event.target.value.trim();

      // Handles e.g. "BRCA1,BRCA2", "BRCA1 BRCA2", and "BRCA1, BRCA2"
      const geneSymbols = searchInput.split(/[, ]/).filter(d => d !== '')
      const geneSymbol = geneSymbols[0];

      urlParams['q'] = geneSymbol;
      updateUrl();

      plotRelatedGenes(geneSymbol);
    }

    function plotGeneFromUrl() {
      if ('q' in urlParams) {
        plotRelatedGenes(urlParams['q']);
      }
    }

    ////
    // End example-specific handling
    //
    // Start generic "Related genes" handling
    ////

    fetchEnsembl = Ideogram.fetchEnsembl;

    function onClickAnnot(annot) {
      document.querySelector('#search-genes').value = annot.name;
      urlParams['q'] = annot.name
      updateUrl();
      plotRelatedGenes(annot.name);
    }

    /**
     * Determines if interaction node might be a gene
     *
     * Some interaction nodes are biological processes; this filters out many.
     * Filtering these out makes downstream queries faster.
     *
     * ixn {Object} Interaction from WikiPathways
     * gene {Object} Gene from MyGene.info
     */
    function maybeGeneSymbol(ixn, gene) {
      return (
        ixn !== '' &&
        !ixn.includes(' ') &&
        !ixn.includes('/') && // e.g. Akt/PKB
        ixn.toLowerCase() !== gene.name.toLowerCase()
      );
    }

    /**
     * Retrieves interacting genes from WikiPathways API
     *
     * Docs:
     * https://webservice.wikipathways.org/ui/
     * https://www.wikipathways.org/index.php/Help:WikiPathways_Webservice/API
     *
     * Examples:
     * https://webservice.wikipathways.org/findInteractions?query=ACE2&format=json
     * https://webservice.wikipathways.org/findInteractions?query=RAD51&format=json
     */
    async function fetchInteractingGenes(gene) {
      let ixns = {};
      let seenNameIds = {}
      const orgNameSimple = ideogram.config.organism.replace(/-/g, ' ');
      const queryString = `?query=${gene.name}&format=json`
      const url =
        `https://webservice.wikipathways.org/findInteractions${queryString}`;

      response = await fetch(url)
      data = await response.json()

      // For each interaction, get nodes immediately upstream and downstream.
      // Filter out pathway nodes that are definitely not gene symbols, then
      // group pathways by (likely) gene symbol. Each interacting gene can have
      // multiple pathways.
      data.result.forEach(interaction => {
        if (interaction.species.toLowerCase() === orgNameSimple) {
          const right = interaction.fields.right.values;
          const left = interaction.fields.left.values;
          const rawIxns = right.concat(left)
          const name = interaction.name;
          const id = interaction.id;

          rawIxns.forEach(rawIxn => {
            const nameId = name + id;
            if (maybeGeneSymbol(rawIxn, gene) && !(nameId in seenNameIds)) {
              seenNameIds[nameId] = 1;
              const ixn = {name: name, pathwayId: id}
              if (rawIxn in ixns) {
                ixns[rawIxn].push(ixn)
              } else {
                ixns[rawIxn] = [ixn]
              }
            }
          })
        }
      })

      return ixns;
    }

    /**
     * Queries MyGene.info API, returns parsed JSON
     *
     * Docs:
     * https://docs.mygene.info/en/v3/
     *
     * Example:
     * https://mygene.info/v3/query?q=symbol:cdk2%20OR%20symbol:brca1&species=9606&fields=symbol,genomic_pos,name
     */
    async function fetchMyGeneInfo(queryString) {
      const myGeneBase = 'https://mygene.info/v3/query';
      let response = await fetch(myGeneBase + queryString);
      let data = await response.json();
      return data;
    }

    /**
     * Summarizes interactions for a gene
     *
     * This comprises most of the content for tooltips for interacting genes.
     */
    function describeInteractions(gene, ixns) {

      let description = gene.name;

      if (typeof ixns !== 'undefined') {
        // ixns is undefined when querying e.g. CDKN1B in human
        const pathwaysBase = 'https://www.wikipathways.org/index.php/Pathway:';
        const links = ixns.map(ixn => {
            const url = `${pathwaysBase}${ixn.pathwayId}`
            return `<a href="${url}" target="_blank">${ixn.name}</a>`;
          }).join('<br/>')

        let pathwayText = 'pathway'
        if (ixns.length > 1) pathwayText += 's'

        description += `
          <br/><br/>
          Interacts in ${pathwayText}:<br/>
          ${links}`;
      }

      return description
    }

    /**
     * Retrieves position and other data on interacting genes from MyGene.info
     */
    async function fetchInteractingGeneAnnots(interactions) {

      let annots = [];
      const geneList = Object.keys(interactions);

      if (geneList.length === 0) return annots

      const ixnParam = geneList.map(ixn => {
        return `symbol:${ixn.trim()}`
      }).join(' OR ')

      const taxid = ideogram.config.taxid;
      const queryString =
        `?q=${ixnParam}&species=${taxid}&fields=symbol,genomic_pos,name`;
      const data = await fetchMyGeneInfo(queryString)

      data.hits.forEach(gene => {

        // If hit lacks position, skip processing
        if ('genomic_pos' in gene === false) return;

        const annot = parseAnnotFromMgiGene(gene, 'purple')
        annots.push(annot)

        const ixns = interactions[gene.symbol];

        const description = describeInteractions(gene, ixns);

        ideogram.annotDescriptions[gene.symbol] = description;
      })

      return annots;
    }

    /**
     * Fetch paralogs of searched gene
     */
     async function fetchParalogPositions(annot, annots) {
      const taxid = ideogram.config.taxid;
      const orgUnderscored = ideogram.config.organism.replace(/[ -]/g, '_');

      const params = `&format=condensed&type=paralogues&target_taxon=${taxid}`;
      let path = `/homology/id/${annot.id}?${params}`
      const ensemblHomologs = await fetchEnsembl(path);
      const homologs = ensemblHomologs.data[0].homologies;

      // Fetch positions of paralogs
      homologIds = homologs.map(homolog => homolog.id)
      path = '/lookup/id/' + orgUnderscored;
      let body = {
        ids: homologIds,
        species: orgUnderscored,
        object_type: 'gene'
      }
      const ensemblHomologGenes = await fetchEnsembl(path, body, 'POST');

      Object.entries(ensemblHomologGenes).map((idGene, i) => {
        let gene = idGene[1]

        // Seen in related genes for SIRT2 in Pan troglodytes
        if ('display_name' in gene === false) return;

        const annot = {
          name: gene.display_name,
          chr: gene.seq_region_name,
          start: gene.start,
          stop: gene.end,
          id: gene.id,
          color: 'pink'
        };

        // Add to start of array, so searched gene gets top z-index
        annots.unshift(annot);
        ideogram.annotDescriptions[annot.name] = gene.description;
      })

      return annots;
    }

    /**
     * Transforms MyGene.info (MGI) gene into Ideogram annotation
     */
    function parseAnnotFromMgiGene(gene, color='red') {
      // Filters out placements on alternative loci scaffolds, an advanced
      // genome assembly feature we are not concerned with in ideograms.
      //
      // Example:
      // https://mygene.info/v3/query?q=symbol:PTPRC&species=9606&fields=symbol,genomic_pos,name
      let genomic_pos = null;
      if (Array.isArray(gene.genomic_pos)) {
        genomic_pos = gene.genomic_pos.filter(pos => {
          return pos.chr in ideogram.chromosomes[ideogram.config.taxid]
        })[0];
      } else {
        genomic_pos = gene.genomic_pos;
      }

      const annot = {
        name: gene.symbol,
        chr: genomic_pos.chr,
        start: genomic_pos.start,
        stop: genomic_pos.end,
        id: genomic_pos.ensemblgene,
        color: color
      };

      return annot
    }

    /**
     * For given gene, finds and draws interacting genes and paralogs
     *
     * @param geneSymbol {String} Gene symbol, e.g. RAD51
     */
    async function plotRelatedGenes(geneSymbol) {

      ideogram.annotDescriptions = {}

      const ideoSel = ideogram.selector
      const annotSel = ideoSel + ' .annot';
      document.querySelectorAll(annotSel).forEach(el => el.remove());

      ideogram.startHideAnnotTooltipTimeout();

      // Refine style
      document.querySelectorAll('.chromosome').forEach(chromosome => {
        chromosome.style.cursor = '';
      })
      const lastChr = [...document.querySelectorAll('.chromosome-set')].slice(-1)[0]
      const left = lastChr.getBoundingClientRect().x + 10;
      const topPx = ideogram.config.chrHeight + 30;
      const style =
        `float: left; position: relative; top: -${topPx}px; left: ${left}px;`;

      // Fetch positon of searched gene
      const taxid = ideogram.config.taxid;
      const queryString =
        `?q=symbol:${geneSymbol}&species=${taxid}&fields=symbol,genomic_pos,name`;
      let data = await fetchMyGeneInfo(queryString)

      const gene = data.hits[0]
      ideogram.annotDescriptions[gene.symbol] = gene.name;

      let annots = [];

      const annot = parseAnnotFromMgiGene(gene);
      annots.push(annot)

      interactions = await fetchInteractingGenes(annot);
      interactingAnnots = await fetchInteractingGeneAnnots(interactions)
      annots = annots.concat(interactingAnnots)

      paralogs = await fetchParalogPositions(annot, annots)

      annots.sort((a, b) => { return b.name.length - a.name.length })

      ideogram.drawAnnots(annots);

      document.querySelector('#_ideogramLegend').style = style;
    }

    /**
     * Enhance tooltip shown on hovering over gene annotation
     */
    function decorateGene(annot) {
      const org = ideogram.getScientificName(ideogram.config.taxid);
      const term = `(${annot.name}[gene])+AND+(${org}[orgn])`;
      const url = `https://ncbi.nlm.nih.gov/gene/?term=${term}`;
      const description = ideogram.annotDescriptions[annot.name].split(' [')[0];
      annot.displayName =
        `<a target="_blank" href="${url}">${annot.name}</a>
        <br/>
        ${description}
        <br/>`;
      return annot
    }

    const shape = 'triangle';

    const legend = [{
      name: '<b>Click gene to search</b>',
      rows: [
        {name: 'Interacting gene', color: 'purple', shape: shape},
        {name: 'Paralogous gene', color: 'pink', shape: shape},
        {name: 'Searched gene', color: 'red', shape: shape}
      ]
    }];

    config = {
      organism: organism,
      container: '#ideogram-container',
      chrWidth: 8,
      chrHeight: 90,
      chrLabelSize: 10,
      annotationHeight: 5,
      showFullyBanded: false,
      rotatable: false,
      legend: legend,
      onClickAnnot: onClickAnnot,
      onWillShowAnnotTooltip: decorateGene
    }

    config.onLoad = plotGeneFromUrl;

    let ideogram = new Ideogram(config);
  </script>
</body>
</html>
